packages/engine/lib/error/src/lib.rs

```diff
@@ -0,0 +1,510 @@
+//! moment with various pros and cons. This crates takes inspiration from an
+//! [RFC to the core library ](https://github.com/nrc/rfcs/blob/dyno/text/0000-dyno.md) to introduce
+//! a reflection-like API - the [`provider`]-API - to add context data to a [`Report`]. Using this,
+//! it becomes possible to attach any data to an error. Beyond this, the design also allows Errors
+//! to have additional requirements, such as _enforcing_ the provision of contextual information by
+//! using [`Report::provide_context()`].
+//!
+//! ### Why not...
+//!
+//! - [`anyhow`]? While the concept is very similar, in `anyhow` you may only provide strings as
+//!   context.
+//! - [`eyre`]? This is a fork of `anyhow`. Some features are nice and also the naming of this crate
+//!   is inspired by `eyre`, however the problem remains the same.
+//! - [`thiserror`]? It "only" provides a macro for implementing [`Error`]. While this can be useful
+//!   for library code, handling errors can be quite tedious as the nesting gets very deep very
+//!   quickly.
+//! - [`snafu`]? Another library for generating errors using a macro. While this library is more
+//!   fully featured than `thiserror`, it is still nesting errors.
+//! - [`failure`]? [`Report`] works similar to `failure::Error`, but also `failure` is only able to
+//!   provide string-like contexts. Also `failure` uses a weird `Fail` trait instead of [`Error`].
+//!
+//! This crates does not claim to be better than the mentioned crates, it's a different approach to
+//! error handling.
+//!
+//! [`Error`]: std::error::Error
+//! [`anyhow`]: https://crates.io/crates/anyhow
+//! [`eyre`]: https://crates.io/crates/eyre
+//! [`thiserror`]: https://crates.io/crates/thiserror
+//! [`snafu`]: https://crates.io/crates/snafu
+//! [`failure`]: https://crates.io/crates/failure
+//! [`Backtrace`]: std::backtrace::Backtrace
+//! [`SpanTrace`]: tracing_error::SpanTrace
+//!
+//! # Usage
+//!
+//! [`Report`] is supposed to be used as the [`Err`] variant of a `Result`. This crates provides a
+//! [`Result<E, C = ()>`] type alias, which uses [`Report<C>`] as [`Err`] variant and can be used as
+//! return type:
+//!
+//! ```
+//! # fn has_permission(_: usize, _: usize) -> bool { true }
+//! # fn get_user() -> Result<usize> { Ok(0) }
+//! # fn get_resource() -> Result<usize> { Ok(0) }
+//! use error::{ensure, Result};
+//!
+//! fn main() -> Result<()> {
+//!     let user = get_user()?;
+//!     let resource = get_resource()?;
+//!
+//!     ensure!(
+//!         has_permission(user, resource),
+//!         "Permission denied for {user} accessing {resource}"
+//!     );
+//!
+//!     # const _: &str = stringify! {
+//!     ...
+//!     # }; Ok(())
+//! }
+//! ```
+//!
+//! A contextual message can be provided to lower level errors.
+//!
+//! ```
+//! use std::collections::HashMap;
+//!
+//! use error::{ensure, error, Result, ResultExt};
+//!
+//! fn lookup_key(map: &HashMap<&str, u64>, key: &str) -> Result<u64> {
```

Comment:
Is there a reason this is in the examples directory _and_ within the actual documentation?

Reply:
It's the module level documentation, so the examples directory is not searched. Also the one in the examples directory is slightly more verbose comment-wise.

Reply:
Fair enough